ocaml/otherlibs/stdlib_alpha/capsule.ml

```diff
@@ -80,9 +105,67 @@ module Mutex = struct
       Password.unsafe_mk ()
 end
 
+module Rwlock = struct
+
+  type 'k t : value mod portable uncontended = { rwlock : Rw.t; mutable poisoned : bool }
+
+  type packed : value mod portable uncontended = P : 'k t -> packed
+
+  exception Poisoned
+
+  let with_write_lock :
```

Comment:
After some further discussion / googling: This is fine.

I referenced the wrong version of POSIX (actually even older than POSIX: my link is to the SUSv2). The current version specifies that write locking a rwlock locked by the same thread is either a EDEADLK or a deadlock, not general UB (and therefore "safe").

Here's the discussion on clarifying this point in POSIX: https://www.austingroupbugs.net/view.php?id=720

(At some point it may be worth testing actual implementations on this point - in particular, the internet suggests that macos may do the wrong thing here)